File: server/cache.ts
import { persistItems, resolveUrl } from './articles.ts';
import { db } from './db.ts';
import { logger } from './logger.ts';
import { parseURL } from './parse-url.ts';
import type { RssItem } from './parse-url.ts';
import type { Feed } from './types.ts';

const log = logger.child({ mod: 'cache' });

export const CACHE_TTL = 5 * 60 * 1000;

// Cap how many feed refreshes run their synchronous work (RSS parse + persist transaction)
// concurrently. Node is single-threaded and better-sqlite3 is synchronous, so an unbounded
// fan-out — today/all-articles calling ensureFresh across every feed, or startup warming
// firing Promise.allSettled over all uncached feeds — would land N parse+persist bursts on the
// event loop at once and stall every concurrent request (even a trivial starred/count). A
// small limit bounds the worst-case block to a couple feeds' worth of work.
const REFRESH_CONCURRENCY = 2;

const setFetchedAt = db.prepare('UPDATE feeds SET last_fetched_at = ? WHERE id = ?');
const feedHasRows = db.prepare('SELECT 1 FROM article_states WHERE feed_id = ? LIMIT 1');

type RefreshResult = { items: RssItem[]; feedName: string };

// ── refresh scheduler + single-flight dedup ──────────────────────────────────────────────
// Every fetch path (on-demand ensureFresh, poller, startup warming) funnels through
// refreshFeed, so both guards live here and cover all callers at once:
//   - single-flight: a refresh already running for a feed is shared, so today + all-articles +
//     poller hitting the same stale feed do ONE fetch, not three.
//   - concurrency limit: at most REFRESH_CONCURRENCY refreshes do their synchronous work at a
//     time, so a page-load/startup fan-out can't bunch every feed's persist into one block.
const inflight = new Map<string, Promise<RefreshResult>>();
let activeRefreshes = 0;
const slotWaiters: Array<() => void> = [];

function acquireRefreshSlot(): Promise<void> {
  if (activeRefreshes < REFRESH_CONCURRENCY) {
    activeRefreshes++;
    return Promise.resolve();
  }
  return new Promise((resolve) => slotWaiters.push(resolve));
}

function releaseRefreshSlot(): void {
  const next = slotWaiters.shift();
  if (next)
    next(); // hand the slot straight to the next waiter (count stays constant)
  else activeRefreshes--;
}

// The single fetch chain: fetch upstream → persist all items into article_states (the durable
// store the list endpoints read from) → stamp the feed's last_fetched_at. Both writes run in
// one transaction. Guarded by a concurrency slot so the synchronous parse+persist never bunches
// (see REFRESH_CONCURRENCY). The caller's AbortSignal is intentionally not threaded here: a
// single-flight refresh is shared, so one caller disconnecting must not abort a fetch others
// are awaiting — parseURL enforces a hard 10s timeout regardless.
async function doRefresh(feed: Feed): Promise<RefreshResult> {
  await acquireRefreshSlot();
  try {
    const parsed = await parseURL(resolveUrl(feed.url));
    const feedName = parsed.title || feed.name;
    db.transaction(() => {
      persistItems(feed, parsed.items, feedName);
      setFetchedAt.run(Date.now(), feed.id);
    })();
    return { items: parsed.items, feedName };
  } finally {
    releaseRefreshSlot();
  }
}

// On-demand reads, background refresh, startup warming and the poller all route through here, so
// every successful fetch persists. Single-flight: concurrent callers for the same feed share one
// in-flight refresh (entry cleared on settle).
export function refreshFeed(feed: Feed): Promise<RefreshResult> {
  const existing = inflight.get(feed.id);
  if (existing) return existing;
  const p = doRefresh(feed).finally(() => inflight.delete(feed.id));
  inflight.set(feed.id, p);
  return p;
}

// Ensure a feed's data is reasonably fresh before its rows are served. Callers read articles
// straight from article_states afterward — this only schedules upstream fetches:
//   - fresh (fetched within TTL): no-op
//   - stale but previously fetched: refresh in the background, serve current rows
//   - never fetched and no rows yet (brand-new feed): await so the first load returns content
//   - never fetched but rows exist (e.g. right after the feed_cache→pub_ts migration):
//     treat as stale, refresh in the background
export async function ensureFresh(feed: Feed): Promise<void> {
  const last = feed.last_fetched_at ?? 0;
  if (last && Date.now() - last < CACHE_TTL) return;
  const backgroundRefresh = () =>
    refreshFeed(feed).catch((err) =>
      log.debug('background refresh failed', { feedId: feed.id, feedUrl: feed.url, err }),
    );
  if (last) {
    backgroundRefresh();
    return;
  }
  if (feedHasRows.get(feed.id)) {
    backgroundRefresh();
  } else {
    await refreshFeed(feed);
  }
}

export let cacheReady = false;

// Warm feeds on startup. Skipped in TEST_DB mode to avoid real network calls. Feeds never
// fetched are warmed up front (and gate cacheReady); already-fetched-but-stale feeds refresh
// in the background.
export function startCacheWarming(): void {
  if (process.env.TEST_DB) return;
  const allFeeds = db.prepare('SELECT * FROM feeds').all() as Feed[];
  const warm = (f: Feed) =>
    refreshFeed(f).catch((err) =>
      log.warn('cache warm failed', { feedId: f.id, feedUrl: f.url, err }),
    );
  // Re-fetch stale (previously fetched) entries in the background.
  allFeeds
    .filter((f) => f.last_fetched_at && Date.now() - f.last_fetched_at >= CACHE_TTL)
    .forEach(warm);
  const uncached = allFeeds.filter((f) => !f.last_fetched_at);
  if (uncached.length === 0) {
    cacheReady = true;
  } else {
    Promise.allSettled(uncached.map(warm)).then(() => {
      cacheReady = true;
      log.info('cache warmed', { feeds: uncached.length });
    });
  }
}

File: server-go/internal/cache/cache.go
// Package cache is the Go port of server/cache.ts: the single fetch chain
// (parse upstream → persist every item into article_states → stamp
// feeds.last_fetched_at) plus the two guards that wrap every fetch path —
// single-flight dedup per feed and a REFRESH_CONCURRENCY slot cap — and the TTL
// scheduler (ensureFresh) callers use before serving a feed's rows.
package cache

import (
	"context"
	"log/slog"
	"sync"
	"time"

	"rss-reader/server-go/internal/crash"
	"rss-reader/server-go/internal/db"
	"rss-reader/server-go/internal/feed"
	"rss-reader/server-go/internal/model"
	"rss-reader/server-go/internal/store"
)

// CacheTTL mirrors CACHE_TTL: a feed fetched within this window is fresh.
const CacheTTL = 5 * time.Minute

// defaultRefreshConcurrency is the fallback REFRESH_CONCURRENCY: at most this many
// refreshes do their fetch+persist work at once, so a page-load / startup / poll
// fan-out can't bunch every feed's persist together. main.go overrides it from
// config via WithConcurrency.
const defaultRefreshConcurrency = 6

// Result is the port of RefreshResult: the parsed items plus the resolved feed
// name (parsed.title || feed.name).
type Result struct {
	Items    []feed.Item
	FeedName string
}

// FetchFunc fetches and parses a feed URL. Injectable so tests can avoid the
// network; production wires feed.ParseURL.
type FetchFunc func(ctx context.Context, url string) (*feed.Parsed, error)

// Cache owns the fetch scheduler state. Unlike the Node module globals it is a
// value so tests get an isolated scheduler + DB.
type Cache struct {
	db    *db.DB
	fetch FetchFunc

	mu       sync.Mutex
	inflight map[string]*flight
	sem      chan struct{} // buffered to refreshConcurrency

	readyMu sync.Mutex
	ready   bool

	// now is the clock (epoch ms); overridable in tests. Defaults to real time.
	now func() int64
}

type flight struct {
	done chan struct{}
	res  Result
	err  error
}

// Option customizes a Cache at construction. Unset options keep the defaults.
type Option func(*Cache)

// WithConcurrency overrides the fetch+persist slot cap (REFRESH_CONCURRENCY).
// n < 1 is ignored so a misconfigured value keeps the default.
func WithConcurrency(n int) Option {
	return func(c *Cache) {
		if n >= 1 {
			c.sem = make(chan struct{}, n)
		}
	}
}

// New builds a Cache over a DB handle. Pass nil fetch to use feed.ParseURL.
func New(handle *db.DB, fetch FetchFunc, opts ...Option) *Cache {
	if fetch == nil {
		fetch = feed.ParseURL
	}
	c := &Cache{
		db:       handle,
		fetch:    fetch,
		inflight: map[string]*flight{},
		sem:      make(chan struct{}, defaultRefreshConcurrency),
		now:      func() int64 { return time.Now().UnixMilli() },
	}
	for _, opt := range opts {
		opt(c)
	}
	return c
}

// RefreshFeed is the single-flight wrapper (port of refreshFeed): concurrent
// callers for the same feed share one in-flight refresh; the entry is cleared on
// settle. Every fetch path (ensureFresh, poller, startup warming) funnels here.
func (c *Cache) RefreshFeed(ctx context.Context, f model.Feed) (Result, error) {
	c.mu.Lock()
	if fl, ok := c.inflight[f.ID]; ok {
		c.mu.Unlock()
		<-fl.done
		return fl.res, fl.err
	}
	fl := &flight{done: make(chan struct{})}
	c.inflight[f.ID] = fl
	c.mu.Unlock()

	fl.res, fl.err = c.doRefresh(ctx, f)

	c.mu.Lock()
	delete(c.inflight, f.ID)
	c.mu.Unlock()
	close(fl.done)
	return fl.res, fl.err
}

// doRefresh is the fetch chain (port of doRefresh): acquire a concurrency slot →
// fetch the (rsshub-resolved) URL → persist every item AND stamp
// last_fetched_at in one transaction. The slot bounds concurrent fetch+persist
// bursts (see refreshConcurrency).
func (c *Cache) doRefresh(ctx context.Context, f model.Feed) (Result, error) {
	c.sem <- struct{}{}
	defer func() { <-c.sem }()

	target, err := store.ResolveURL(c.db.Reader(), f.URL)
	if err != nil {
		return Result{}, err
	}
	parsed, err := c.fetch(ctx, target)
	if err != nil {
		return Result{}, err
	}
	feedName := f.Name
	if parsed.Title != "" {
		feedName = parsed.Title
	}
	now := c.now()
	if err := store.RefreshPersist(c.db.Writer(), f.ID, feedName, f.URL, parsed.Items, now); err != nil {
		return Result{}, err
	}
	return Result{Items: parsed.Items, FeedName: feedName}, nil
}

// EnsureFresh is the port of ensureFresh: schedule an upstream fetch as needed
// before a feed's rows are served (callers read article_states afterward).
//   - fresh (fetched within TTL): no-op
//   - stale but previously fetched: refresh in the background, serve current rows
//   - never fetched, no rows yet (brand-new feed): await so first load has content
//   - never fetched but rows exist: treat as stale, refresh in the background
func (c *Cache) EnsureFresh(ctx context.Context, f model.Feed) error {
	var last int64
	if f.LastFetchedAt != nil {
		last = *f.LastFetchedAt
	}
	if last != 0 && c.now()-last < CacheTTL.Milliseconds() {
		return nil
	}
	if last != 0 {
		c.backgroundRefresh(f)
		return nil
	}
	hasRows, err := store.FeedHasRows(c.db.Reader(), f.ID)
	if err != nil {
		return err
	}
	if hasRows {
		c.backgroundRefresh(f)
		return nil
	}
	_, err = c.RefreshFeed(ctx, f)
	return err
}

// backgroundRefresh fires a refresh without blocking the caller; errors are
// swallowed (the Node path logs at debug and moves on).
func (c *Cache) backgroundRefresh(f model.Feed) {
	go func() {
		defer crash.Recover(slog.Default(), "background-refresh")
		_, _ = c.RefreshFeed(context.Background(), f)
	}()
}

// Ready reports whether startup cache warming has completed (gates readiness).
func (c *Cache) Ready() bool {
	c.readyMu.Lock()
	defer c.readyMu.Unlock()
	return c.ready
}

func (c *Cache) setReady() {
	c.readyMu.Lock()
	c.ready = true
	c.readyMu.Unlock()
}

// StartCacheWarming is the port of startCacheWarming: warm never-fetched feeds up
// front (gating Ready), refresh already-fetched-but-stale feeds in the
// background. It returns immediately; Ready() flips true once the uncached feeds
// settle. Callers gate it on TEST mode themselves (Node keys off TEST_DB).
func (c *Cache) StartCacheWarming() error {
	feeds, err := store.ListFeeds(c.db.Reader())
	if err != nil {
		return err
	}
	var uncached []model.Feed
	now := c.now()
	for _, f := range feeds {
		if f.LastFetchedAt != nil && now-*f.LastFetchedAt >= CacheTTL.Milliseconds() {
			c.backgroundRefresh(f) // stale: refresh in the background
		} else if f.LastFetchedAt == nil {
			uncached = append(uncached, f)
		}
	}
	if len(uncached) == 0 {
		c.setReady()
		return nil
	}
	go func() {
		defer crash.Recover(slog.Default(), "cache-warming")
		var wg sync.WaitGroup
		for _, f := range uncached {
			wg.Add(1)
			go func(f model.Feed) {
				defer crash.Recover(slog.Default(), "cache-warming-feed")
				defer wg.Done()
				_, _ = c.RefreshFeed(context.Background(), f)
			}(f)
		}
		wg.Wait()
		c.setReady()
	}()
	return nil
}
